### ISSUE TITLE
Use Karma to run headless unit tests

### ISSUE THREAD

**Issue by mxstbr** [mxstbr/react-boilerplate#79]
Currently we don't compile our tests with webpack, which means any component we test that imports non-js/non-css files (e.g. images) will not work.

Karma fixes this because it uses webpack and renders against a headless browser, which means if our application works with the webpack config it'll work in the tests.

Example Karma config: https://github.com/jquense/react-widgets/blob/master/karma.conf.js
Run with: `karma start karma.conf.js --single-run=true`

> Note: If you submit a PR, do it against the `v3.0.0` branch please!

**Comment by mxstbr** [comment]
As I've never before working with Karma, this has proven to be harder than I initially thought.

If somebody else wants to take a stab at this please do so, I really want this to happen!

**Comment by jbinto** [comment]
I was able to get the example Karma config linked in the issue working for one of my small projects.

It is not headless: on my OS X dev machine a Chrome window will pop up. It does work on Travis though (via `xvfb`), so that's good enough for me.

I'll send a PR shortly.

**Comment by mxstbr** [comment]
@jbinto I'd love a PR, thank you so much! I thought you could use PhatomJS instead of a browser, and that'd be headless?

**Comment by jbinto** [comment]
From what I've cobbled together from google, it doesn't look like PhantomJS is the way to go anymore. 
PhantomJS is based on old (c.2013) Webkit code, and the engines are different enough to cause trouble.

See e.g. 

* http://www.gizra.com/content/phantomjs-chrome-docker-selenium-standalone/
* http://stackoverflow.com/a/29433038/19779

and my favourite:

* https://twitter.com/stevenlangbroek/status/676445439940513793 
* https://gist.github.com/StevenLangbroek/552f639c95e2cf9f74ad

Many projects have moved to Chrome, e.g. [react-router](https://github.com/rackt/react-router/blame/af6c1ef9856267b223f666474171e62b1229af3b/karma.conf.js#L88). 

I also resisted moving away from jsdom/mocha, since it's so light-weight. But in the end I think Chrome makes more sense. 

--


As for a PR: I'm having some trouble with the webpack config right now. I think for some reason, it's running all of `node_modules` through webpack. 

I'll try to sort it out now.

**Comment by mxstbr** [comment]
Amazing, thanks so much for the insight and the sources. 👏

I had similar problems with webpack, so I'm really looking forward to seeing how you solve them!

**Comment by jbinto** [comment]
Okay, so I have karma working for the most part:

https://github.com/jbinto/react-boilerplate/commit/83a1d1d2ee33acee66305919a2a9f81e502b5507

When I said "everything gets run through webpack", at first I thought that was somehow wrong, but now of course it makes sense: it has to bundle up all of the dev dependencies in order to run the tests in the browser. There is so much output, it was hard to see actual errors. 

I had to replace `rewire` with `babel-plugin-rewire` due to incompatibilities with karma. This is working fine.

--

But getting `chai-enzyme` to work is problematic. Apparently, you can't just call `chai.use()` inside of Karma. There's a whole "plugin" ecosystem. And there's no Karma plugin for `chai-enzyme`. 

So I tried to write one. But I don't really know what I'm doing, and I've hit a wall.

Another chai plugin I've used is `dirty-chai`. (It requires you to put `()` at the end of a `to.be.true` style call, so you don't have to reconfigure your linter.) There's a `karma-dirty-chai` package, which is just a bit of glue for karma, so I tried to adapt a `karma-chai-enzyme` from that.

It's not that simple. All of the karma plugins I can find are plain ES5, and require other ES5/CommonJS code. `chai-enzyme` is published to npm as an ES6 module (it's precompiled, but that doesn't seem to matter). 

When I try [my copy-pasted karma plugin implementation](https://github.com/jbinto/karma-chai-enzyme/blob/master/index.js) that tells Karma to require `chai-enzyme` I get 

```
  Uncaught ReferenceError: exports is not defined
  at /Users/jbinto/dev/react-boilerplate/node_modules/chai-enzyme/build/index.js:3
```


--

This is how I understand the situation (I may be wrong at any point here):

Karma runs, calls webpack and makes a bundle including all of the test/devDependency code. webpack  runs everything through babel. The bundle is now generated. Karma executes the bundle, including my plugin code, which tells it to require `chai-enzyme`. [The npm published chai-enzyme is an ES6 module](https://npmcdn.com/chai-enzyme@0.2.2). The environment we're executing in at this point (plain node on the webpack bundle) doesn't understand ES6 modules. 

Gonna take a bit of a break on this.

**Comment by mxstbr** [comment]
Wow, that's a great breakdown. I'm going to loop in the creator of chai-enzyme here – @vesln any input on this?

We only very recently introduced chai-enzyme, see #104. If this is too big of an issue to work around of, we can easily revert back to the testing state before chai-enzyme. What do you reckon, should we do that?

**Comment by jbinto** [comment]
I'd be ok with reverting that and calling it a day, but I do see the value in `chai-enzyme` and would like to get it all working together at some point.

Aside: The whole plugin thing leaves a bad taste in my mouth compared to the simplicity of `chai.use`. It's one of the things that makes Karma feel heavier. I mean, you can call `chai.use()` multiple times to include multiple Chai plugins. But if you want that in Karma, it seems you now need N*M packages for `karma-foo-bar`, `karma-foo-bar-baz`, etc.

**Comment by jbinto** [comment]
OK, I reverted https://github.com/mxstbr/react-boilerplate/commit/305cb56ff97fd8516871fbc62ab2f33c182fa7b6 for now and everything is working with Karma:

I noticed that Travis uses an old version of Chromium. I found some instructions for Travis to install Chrome stable from a `.deb`, and that works well.

Everything passing on Chrome 48.0.2564:

https://travis-ci.org/jbinto/react-boilerplate/builds/104545478

I'll send the PR now.
